featurebyte/service/validator/entity_relationship_validator.py

```diff
@@ -0,0 +1,122 @@
+        self.entity_service = entity_service
+        self.id_to_ancestors: Dict[ObjectId, Set[AncestorData]] = defaultdict(set)
+
+    def reset(self) -> None:
+        """
+        Reset the validator
+        """
+        self.id_to_ancestors = defaultdict(set)
+
+    def _update_ancestor_mapping(
+        self, relationship: EntityRelationshipInfo, feature_name: str
+    ) -> None:
+        """
+        Add ancestor IDs to the ID to ancestor IDs mapping
+
+        Parameters
+        ----------
+        relationship: EntityRelationshipInfo
+            Relationship to add
+        feature_name: str
+            Feature name of the relationship
+        """
+        if relationship.relationship_type == RelationshipType.CHILD_PARENT:
+            entity_id = relationship.entity_id
+            parent_entity_id = relationship.related_entity_id
+            self.id_to_ancestors[entity_id].add(AncestorData(parent_entity_id, (feature_name,)))
+            for ancestor in self.id_to_ancestors[parent_entity_id]:
+                self.id_to_ancestors[entity_id].add(
+                    AncestorData(
+                        ancestor_id=ancestor.ancestor_id,
+                        feature_names=ancestor.feature_names + (feature_name,),
+                    )
+                )
+
+    async def _validate_relationship(
+        self, relationship: EntityRelationshipInfo, feature_name: str
+    ) -> None:
+        """
+        Validate relationship by checking if it conflicts with existing relationships
+
+        Parameters
+        ----------
+        relationship: EntityRelationshipInfo
+            Relationship to add
+        feature_name: str
+            Feature name of the relationship
+
+        Raises
+        ------
+        EntityRelationshipConflictError
+            If the relationship conflicts with an existing relationship
+        """
+        if relationship.relationship_type == RelationshipType.CHILD_PARENT:
+            # check if the relationship conflicts with existing relationships
+            parent_ancestors = self.id_to_ancestors[relationship.related_entity_id]
+            for ancestor in parent_ancestors:
+                if ancestor.ancestor_id == relationship.entity_id:
+                    feature_names = sorted(set(ancestor.feature_names))
+                    entity = await self.entity_service.get_document(
+                        document_id=relationship.entity_id
+                    )
+                    related_entity = await self.entity_service.get_document(
+                        document_id=relationship.related_entity_id
+                    )
+                    raise EntityRelationshipConflictError(
+                        f"Entity '{entity.name}' is an ancestor of "
+                        f"'{related_entity.name}' (based on features: {feature_names}) "
+                        f"but feature '{feature_name}' has a child-parent relationship between them."
```

Comment:
Would it be clearer if the message is something like  this?

> ... but this relationship is inverted based on '{feature_name}'. Consider excluding '{feature_name}' from the Feature List.

Reply:
Google barb complaints about the `inverted` as it can be confusing for some users. How about
```
                        f"Entity '{entity.name}' is an ancestor of "
                        f"'{related_entity.name}' (based on features: {ancestor_feature_names}) "
                        f"but '{entity.name} is a child of '{related_entity.name}' based on '{feature_name}'. "
                        f"Consider excluding '{feature_name}' from the Feature List."
```

Reply:
Yes, I think mentioning which is child and which is parent is better 👍